crates/sequencer/src/deposit_data_mempool.rs

```
@@ -42,30 +45,65 @@ impl DepositDataMempool {
     ///
     /// # Arguments
     /// * `limit_per_block` - Maximum number of deposits to return
     ///
     /// # Returns
     /// A vector of deposit transaction data, limited by the specified amount
-    pub fn fetch_deposits(&mut self, limit_per_block: usize) -> Vec<Vec<u8>> {
+    pub fn fetch_deposits(&mut self, limit_per_block: usize) -> Vec<Deposit> {
         let number_of_deposits = self.accepted_deposit_txs.len().min(limit_per_block);
         SM.deposit_data_mempool_txs
             .set(self.accepted_deposit_txs.len() as f64);
-        let deposits: Vec<Vec<u8>> = self
+        let deposits: Vec<Deposit> = self
             .accepted_deposit_txs
-            .drain(..number_of_deposits)
+            .iter()
+            .take(number_of_deposits)
+            .cloned()
             .collect();
 
-        // Remove fetched deposits from the pending set
-        for deposit in &deposits {
+        deposits
+    }
+
+    /// Removes specific deposits from the mempool after they have been successfully included in a block
+    ///
+    /// # Arguments
+    /// * `deposits_to_remove` - The deposits that were successfully included
+    ///
+    /// # Returns
+    /// The number of deposits actually removed
+    #[instrument(level = "trace", skip_all, ret)]
+    pub fn remove_deposits(&mut self, deposits_to_remove: &[Deposit]) -> usize {
+        let mut removed_count = 0;
+
+        // Calculate txids for the deposits to remove
+        let mut txids_to_remove = HashSet::new();
+        for deposit in deposits_to_remove {
             if let Ok(txid) = Self::calc_tx_id(deposit) {
-                self.pending_deposits.remove(txid.as_slice());
+                txids_to_remove.insert(txid.to_vec());
             }
         }
 
-        deposits
+        // Retain only deposits that are not in the removal set
+        self.accepted_deposit_txs.retain(|deposit| {
+            if let Ok(txid) = Self::calc_tx_id(deposit) {
```

Shouldn't we return the error here as that would only happen in a corrupted state ? Above as well

---

yeah, good point.

@rakanalh we can panic on calc_tx_id errors in this function. it should never be called with non-deposit Vec

---

I checked. `calc_tx_id` doesn't really return an error... it always returns `Ok`. We should change the signature of the method.